### PR TITLE
Add `CollectionAssert` hierarchy

### PR DESCRIPTION
**Pull request by scordio**
Prerequisite for #2102.

### CODE REVIEW

**Comment by scordio** [comment]
@joel-costigliola could you please check it to make sure I haven't missed anything?

File: src/main/java/org/assertj/core/api/ClassBasedNavigableCollectionAssert.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import java.lang.reflect.Constructor;
+import java.util.Collection;
+
+/**
+ * Build the Assert instance by reflection.
+ * 
+ * @since 3.21.0
+ */
+//@format:off
+public class ClassBasedNavigableCollectionAssert<SELF extends ClassBasedNavigableCollectionAssert<SELF, ACTUAL, ELEMENT, ELEMENT_ASSERT>,
```
Comment (scordio) [inline src/main/java/org/assertj/core/api/ClassBasedNavigableCollectionAssert.java:26]: I'm not sure about the concrete use cases for `ClassBasedNavigableCollectionAssert` and `FactoryBasedNavigableCollectionAssert`, and if it makes sense at all to have them. For now, the structure is similar to `ListAssert` and `IterableAssert`.

Comment (joel-costigliola) [reply src/main/java/org/assertj/core/api/ClassBasedNavigableCollectionAssert.java:26]: it's to get narrowed type assertions when navigating to elements,ex:
```java
Iterable<String> hobbits = newHashSet("frodo", "sam", "pippin");
assertThat(hobbits, StringAssert.class).first()
                                        .startsWith("fro")
                                        .endsWith("do");
```
This was introduced before instanceof factories which fills a similar purpose.
I don't mind leaving them out if that's possible.

File: src/test/java/org/assertj/core/api/InstanceOfAssertFactoriesTest.java [hunk 2]
```diff
@@ -1089,6 +1092,26 @@ void typed_iterator_factory_should_allow_typed_iterator_assertions() {
     result.hasNext();
   }
 
+  @Test
+  void collection_factory_should_allow_collection_assertions() {
+    // GIVEN
+    Object value = asList("Homer", "Marge", "Bart", "Lisa", "Maggie");
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/InstanceOfAssertFactoriesTest.java:1098]: minor: I usually use `list()` from `org.assertj.core.util.util.Lists`

File: src/main/java/org/assertj/core/api/ClassBasedNavigableCollectionAssert.java (hunk 1 above)
Comment (scordio) [reply src/main/java/org/assertj/core/api/ClassBasedNavigableCollectionAssert.java:26]: Ok, so I'll keep it simple and remove them. We can always add them later if users request them.

**Pull request merged**